datafusion/catalog/src/async.rs

```diff
@@ -0,0 +1,747 @@
+}
+impl ResolvedCatalogProviderBuilder {
+    fn new(async_provider: Arc<dyn AsyncCatalogProvider>) -> Self {
+        Self {
+            cached_schemas: HashMap::new(),
+            async_provider,
+        }
+    }
+    fn finish(self) -> Arc<dyn CatalogProvider> {
+        let cached_schemas = self
+            .cached_schemas
+            .into_iter()
+            .filter_map(|(key, maybe_value)| {
+                maybe_value.map(|value| (key, value.finish()))
+            })
+            .collect();
+        Arc::new(ResolvedCatalogProvider { cached_schemas })
+    }
+}
+
+/// A catalog provider list that looks up catalogs in a cache
+///
+/// Instances are created by the [`AsyncCatalogProviderList::resolve`] method
+#[derive(Debug)]
+struct ResolvedCatalogProviderList {
+    cached_catalogs: HashMap<String, Arc<dyn CatalogProvider>>,
+}
+impl CatalogProviderList for ResolvedCatalogProviderList {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn register_catalog(
+        &self,
+        _name: String,
+        _catalog: Arc<dyn CatalogProvider>,
+    ) -> Option<Arc<dyn CatalogProvider>> {
+        unimplemented!("resolved providers cannot handle registration APIs")
+    }
+
+    fn catalog_names(&self) -> Vec<String> {
+        self.cached_catalogs.keys().cloned().collect()
+    }
+
+    fn catalog(&self, name: &str) -> Option<Arc<dyn CatalogProvider>> {
+        self.cached_catalogs.get(name).cloned()
+    }
+}
+
+/// A trait for schema providers that must resolve tables asynchronously
+///
+/// The [`SchemaProvider::table`] method _is_ asynchronous.  However, this is primarily for convenience and
+/// it is not a good idea for this method to be slow as this will cause poor planning performance.
+///
+/// It is a better idea to resolve the tables once and cache them in memory for the duration of
+/// planning.  This trait helps implement that pattern.
+///
+/// After implementing this trait you can call the [`AsyncSchemaProvider::resolve`] method to get an
+/// `Arc<dyn SchemaProvider>` that contains a cached copy of the referenced tables.  The `resolve`
+/// method can be slow and asynchronous as it is only called once, before planning.
+#[async_trait]
+pub trait AsyncSchemaProvider: Send + Sync {
+    /// Lookup a table in the schema provider
+    async fn table(&self, name: &str) -> Result<Option<Arc<dyn TableProvider>>>;
+    /// Creates a cached provider that can be used to execute a query containing given references
+    ///
+    /// This method will walk through the references and look them up once, creating a cache of table
+    /// providers.  This cache will be returned as a synchronous TableProvider that can be used to plan
+    /// and execute a query containing the given references.
+    ///
+    /// This cache is intended to be short-lived for the execution of a single query.  There is no mechanism
+    /// for refresh or eviction of stale entries.
```

Comment:
maybe we can add a reference to the example, like this (can be done as a follow on PR as well)

```suggestion
    /// for refresh or eviction of stale entries.
    ////
    /// See the [remote_catalog.rs] for a full end to end example
    ///
    /// [remote_catalog.rs]: https://github.com/apache/datafusion/blob/main/datafusion-examples/examples/remote_catalog.rs
```

Reply:
Follow on PR to add doc: 
- https://github.com/apache/datafusion/pull/14062